### PR TITLE
generate crds and mappers for DiscoveryEngineEngine

### PR DESCRIPTION
**Pull request by justinsb**
- **tests: create generic fuzz tester**
- **discoveryengine: initial generation of CRDs**
- **discovery: generate mappers**
- **autogen: generate apimachinery**
- **mappers: tweaks to pass fuzz tests**
- **chore: controllerbuilder better handling for enum slices**
- **chore: enhanced fuzzing support for enum lists and bytes**

### CODE REVIEW

**Review by yuwenma — COMMENTED** [review]
The two commits related to API and mapper looks good.

File: apis/discoveryengine/v1alpha1/engine_reference.go [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+	// Should be in the format "projects/<projectID>/locations/<location>/engines/<engineID>".
+	External string `json:"external,omitempty"`
+
+	// The name of a DiscoveryEngineEngine resource.
+	Name string `json:"name,omitempty"`
+
+	// The namespace of a DiscoveryEngineEngine resource.
+	Namespace string `json:"namespace,omitempty"`
+
+	parent *DiscoveryEngineEngineParent
```
Comment (yuwenma) [inline apis/discoveryengine/v1alpha1/engine_reference.go:45]: My bad, but I think we want to fix this.

File: apis/discoveryengine/v1alpha1/engine_types.go [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+)
+
+var DiscoveryEngineEngineGVK = GroupVersion.WithKind("DiscoveryEngineEngine")
+
+// DiscoveryEngineEngineSpec defines the desired state of DiscoveryEngineEngine
+// +kcc:proto=google.cloud.discoveryengine.v1.Engine
+type DiscoveryEngineEngineSpec struct {
+	// Required. The display name of the engine. Should be human readable. UTF-8
+	// encoded string with limit of 1024 characters.
+	DisplayName *string `json:"displayName,omitempty"`
```
Comment (yuwenma) [inline apis/discoveryengine/v1alpha1/engine_types.go:30]: Just to double check, Is this a required field?

File: apis/discoveryengine/v1alpha1/engine_types.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+	// The data stores associated with this engine.
+	// For SOLUTION_TYPE_SEARCH and SOLUTION_TYPE_RECOMMENDATION type of
+	// engines, they can only associate with at most one data store.
+	// If solution_type is SOLUTION_TYPE_CHAT, multiple DataStores in the
+	// same Collection can be associated here.
+	// Note that when used in CreateEngineRequest, one DataStore id must be
+	// provided as the system will use it for necessary initializations.
+	DataStoreIds []string `json:"dataStoreIds,omitempty"`
```
Comment (yuwenma) [inline apis/discoveryengine/v1alpha1/engine_types.go:39]: This seems to be a reference.

File: apis/discoveryengine/v1alpha1/engine_types.go [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+	// Configurations for the Search Engine. Only applicable if
+	// solution_type is SOLUTION_TYPE_SEARCH.
+	SearchEngineConfig *Engine_SearchEngineConfig `json:"searchEngineConfig,omitempty"`
+
+	/* Immutable. The Project that this resource belongs to. */
+	ProjectRef *refs.ProjectRef `json:"projectRef"`
+
+	/* Immutable. Location of the resource. */
+	Location string `json:"location"`
```
Comment (yuwenma) [inline apis/discoveryengine/v1alpha1/engine_types.go:71]: Suggest adding the `+required` tag

File: apis/discoveryengine/v1alpha1/engine_types.go [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="ResourceID field is immutable"
```
Comment (yuwenma) [inline apis/discoveryengine/v1alpha1/engine_types.go:73]: mind removing this? We will fix the api template.

File: apis/discoveryengine/v1alpha1/engine_types.go (hunk 4 above)
Comment (justinsb) [reply apis/discoveryengine/v1alpha1/engine_types.go:71]: Good call - done!

File: apis/discoveryengine/v1alpha1/engine_types.go (hunk 2 above)
Comment (justinsb) [reply apis/discoveryengine/v1alpha1/engine_types.go:30]: Good call - added 	// +required

File: apis/discoveryengine/v1alpha1/engine_types.go (hunk 5 above)
Comment (justinsb) [reply apis/discoveryengine/v1alpha1/engine_types.go:85]: Sure.  Is there a rule though - e.g. should location have the CEL expression for immutability?

File: apis/discoveryengine/v1alpha1/engine_types.go (hunk 3 above)
Comment (justinsb) [reply apis/discoveryengine/v1alpha1/engine_types.go]: Converted to a ref

File: apis/discoveryengine/v1alpha1/engine_reference.go (hunk 1 above)
Comment (justinsb) [reply apis/discoveryengine/v1alpha1/engine_reference.go]: Done!

Comment (justinsb) [reply apis/discoveryengine/v1alpha1/engine_reference.go]: (And no worries!)

File: apis/discoveryengine/v1alpha1/engine_types.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	/* Immutable. The Project that this resource belongs to. */
+	ProjectRef *refs.ProjectRef `json:"projectRef"`
+
+	/* Immutable. Location of the resource. */
+	// +required
+	Location string `json:"location"`
+
+	// Immutable. The collection for the Engine.
+	// +required
+	Collection string `json:"collection"`
```
Comment (yuwenma) [inline apis/discoveryengine/v1alpha1/engine_types.go:82]: This really looks like a resource reference. But it is not. Do you know what happens?

**Review by yuwenma — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**